File: my-reader-mobile/modules/readium/ios/Search/SearchSessionStore.swift
import Foundation
import ReadiumShared

/// Pending creation of a native search iterator.
final class PendingSearchRequest: @unchecked Sendable {
  let id = UUID().uuidString
  let publicationId: String

  private let lock = NSLock()
  private var task: Task<Void, Never>?
  private var cancelled = false

  init(publicationId: String) {
    self.publicationId = publicationId
  }

  func attach(task: Task<Void, Never>) {
    lock.lock()
    if cancelled {
      lock.unlock()
      task.cancel()
      return
    }
    self.task = task
    lock.unlock()
  }

  func cancel() {
    lock.lock()
    guard !cancelled else {
      lock.unlock()
      return
    }
    cancelled = true
    let task = task
    self.task = nil
    lock.unlock()

    task?.cancel()
  }
}

/// One stateful REP-007 iterator exposed to JavaScript by an opaque id.
final class NativeSearchSession: @unchecked Sendable {
  let id = UUID().uuidString
  let publicationId: String
  let iterator: any SearchIterator

  private let lock = NSLock()
  private var nextTask: Task<Void, Never>?
  private var closed = false

  init(publicationId: String, iterator: any SearchIterator) {
    self.publicationId = publicationId
    self.iterator = iterator
  }

  /// Starts at most one `next()` call at a time for this iterator.
  func startNext(_ operation: @escaping @Sendable () async -> Void) -> Bool {
    lock.lock()
    guard !closed, nextTask == nil else {
      lock.unlock()
      return false
    }
    nextTask = Task { await operation() }
    lock.unlock()
    return true
  }

  func finishNext() {
    lock.lock()
    nextTask = nil
    lock.unlock()
  }

  /// Runs the delivery while cancellation is excluded by the session lock.
  func deliverIfActive(_ delivery: () -> Void) -> Bool {
    lock.lock()
    guard !closed else {
      lock.unlock()
      return false
    }
    delivery()
    lock.unlock()
    return true
  }

  /// Marks the iterator exhausted while delivering its final page.
  func complete(_ delivery: () -> Void) -> Bool {
    lock.lock()
    guard !closed else {
      lock.unlock()
      return false
    }
    closed = true
    delivery()
    lock.unlock()

    iterator.close()
    return true
  }

  func cancel() {
    lock.lock()
    guard !closed else {
      lock.unlock()
      return
    }
    closed = true
    let task = nextTask
    nextTask = nil
    lock.unlock()

    task?.cancel()
    iterator.close()
  }
}

/// Thread-safe ownership and lifecycle for pending and active searches.
final class SearchSessionStore: @unchecked Sendable {
  static let shared = SearchSessionStore()

  private let queue = DispatchQueue(label: "my-reader.readium.search-session-store")
  private var pendingByPublication: [String: PendingSearchRequest] = [:]
  private var sessionsById: [String: NativeSearchSession] = [:]

  private init() {}

  /// Starting a new query invalidates every older query for the publication.
  func begin(publicationId: String) -> PendingSearchRequest {
    queue.sync {
      pendingByPublication.removeValue(forKey: publicationId)?.cancel()

      let staleIds = sessionsById.values
        .filter { $0.publicationId == publicationId }
        .map(\.id)
      for id in staleIds {
        sessionsById.removeValue(forKey: id)?.cancel()
      }

      let pending = PendingSearchRequest(publicationId: publicationId)
      pendingByPublication[publicationId] = pending
      return pending
    }
  }

  func install(
    _ iterator: any SearchIterator,
    for pending: PendingSearchRequest
  ) -> NativeSearchSession? {
    queue.sync {
      guard pendingByPublication[pending.publicationId] === pending else {
        return nil
      }
      pendingByPublication.removeValue(forKey: pending.publicationId)

      let session = NativeSearchSession(
        publicationId: pending.publicationId,
        iterator: iterator
      )
      sessionsById[session.id] = session
      return session
    }
  }

  /// Removes a pending request only when it is still the current query.
  @discardableResult
  func discard(_ pending: PendingSearchRequest) -> Bool {
    queue.sync {
      guard pendingByPublication[pending.publicationId] === pending else {
        return false
      }
      pendingByPublication.removeValue(forKey: pending.publicationId)
      return true
    }
  }

  func session(id: String) -> NativeSearchSession? {
    queue.sync { sessionsById[id] }
  }

  func remove(_ session: NativeSearchSession) {
    queue.sync {
      guard sessionsById[session.id] === session else { return }
      sessionsById.removeValue(forKey: session.id)
    }
  }

  /// Idempotent cancellation by the opaque JavaScript session id.
  func cancel(sessionId: String) {
    queue.sync {
      sessionsById.removeValue(forKey: sessionId)?.cancel()
    }
  }

  func cancel(publicationId: String) {
    queue.sync {
      pendingByPublication.removeValue(forKey: publicationId)?.cancel()

      let staleIds = sessionsById.values
        .filter { $0.publicationId == publicationId }
        .map(\.id)
      for id in staleIds {
        sessionsById.removeValue(forKey: id)?.cancel()
      }
    }
  }

  func cancelAll() {
    queue.sync {
      pendingByPublication.values.forEach { $0.cancel() }
      sessionsById.values.forEach { $0.cancel() }
      pendingByPublication.removeAll()
      sessionsById.removeAll()
    }
  }
}

File: my-reader-mobile/modules/readium/android/src/main/java/com/myreader/readium/Search/SearchSessionStore.kt
package com.myreader.readium.Search

import java.util.UUID
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import org.readium.r2.shared.ExperimentalReadiumApi
import org.readium.r2.shared.publication.services.search.SearchIterator

internal class PendingSearchRequest(
  val publicationId: String,
) {
  val id: String = UUID.randomUUID().toString()

  private val lock = Any()
  private var job: Job? = null
  private var cancelled = false

  fun attach(job: Job) {
    val cancelNow = synchronized(lock) {
      if (cancelled) {
        true
      } else {
        this.job = job
        false
      }
    }
    if (cancelNow) job.cancel()
  }

  fun cancel() {
    val activeJob = synchronized(lock) {
      if (cancelled) return
      cancelled = true
      job.also { job = null }
    }
    activeJob?.cancel()
  }
}
@OptIn(ExperimentalReadiumApi::class)
internal class NativeSearchSession(
  val publicationId: String,
  val iterator: SearchIterator,
) {
  val id: String = UUID.randomUUID().toString()

  private val lock = Any()
  private var nextJob: Job? = null
  private var closed = false

  /** Starts at most one iterator request at a time. */
  fun startNext(scope: CoroutineScope, operation: suspend () -> Unit): Boolean {
    val job = synchronized(lock) {
      if (closed || nextJob != null) return false
      scope.launch(start = CoroutineStart.LAZY) { operation() }
        .also { nextJob = it }
    }
    job.start()
    return true
  }

  fun finishNext() {
    synchronized(lock) { nextJob = null }
  }

  /** Runs result delivery while cancellation is excluded by the session lock. */
  fun deliverIfActive(delivery: () -> Unit): Boolean = synchronized(lock) {
    if (closed) return false
    delivery()
    true
  }

  /** Marks the iterator exhausted while delivering its final page. */
  fun complete(delivery: () -> Unit): Boolean {
    val completed = synchronized(lock) {
      if (closed) return false
      closed = true
      delivery()
      true
    }
    if (completed) iterator.close()
    return completed
  }

  fun cancel() {
    val activeJob = synchronized(lock) {
      if (closed) return
      closed = true
      nextJob.also { nextJob = null }
    }
    activeJob?.cancel()
    iterator.close()
  }
}

@OptIn(ExperimentalReadiumApi::class)
internal object SearchSessionStore {
  private val lock = Any()
  private val pendingByPublication = mutableMapOf<String, PendingSearchRequest>()
  private val sessionsById = mutableMapOf<String, NativeSearchSession>()

  /** Starting a new query invalidates every older query for the publication. */
  fun begin(publicationId: String): PendingSearchRequest = synchronized(lock) {
    pendingByPublication.remove(publicationId)?.cancel()

    val staleIds = sessionsById.values
      .filter { it.publicationId == publicationId }
      .map { it.id }
    staleIds.forEach { id -> sessionsById.remove(id)?.cancel() }

    PendingSearchRequest(publicationId).also {
      pendingByPublication[publicationId] = it
    }
  }

  fun install(
    iterator: SearchIterator,
    pending: PendingSearchRequest,
  ): NativeSearchSession? = synchronized(lock) {
    if (pendingByPublication[pending.publicationId] !== pending) return null
    pendingByPublication.remove(pending.publicationId)

    NativeSearchSession(pending.publicationId, iterator).also {
      sessionsById[it.id] = it
    }
  }

  /** Removes a pending request only when it is still the current query. */
  fun discard(pending: PendingSearchRequest): Boolean = synchronized(lock) {
    if (pendingByPublication[pending.publicationId] !== pending) return false
    pendingByPublication.remove(pending.publicationId)
    true
  }

  fun session(id: String): NativeSearchSession? = synchronized(lock) {
    sessionsById[id]
  }

  fun remove(session: NativeSearchSession) {
    synchronized(lock) {
      if (sessionsById[session.id] === session) {
        sessionsById.remove(session.id)
      }
    }
  }

  /** Idempotent cancellation by the opaque JavaScript session id. */
  fun cancel(sessionId: String) {
    synchronized(lock) { sessionsById.remove(sessionId)?.cancel() }
  }

  fun cancelPublication(publicationId: String) {
    synchronized(lock) {
      pendingByPublication.remove(publicationId)?.cancel()

      val staleIds = sessionsById.values
        .filter { it.publicationId == publicationId }
        .map { it.id }
      staleIds.forEach { id -> sessionsById.remove(id)?.cancel() }
    }
  }

  fun cancelAll() {
    synchronized(lock) {
      pendingByPublication.values.forEach { it.cancel() }
      sessionsById.values.forEach { it.cancel() }
      pendingByPublication.clear()
      sessionsById.clear()
    }
  }
}
